### PR TITLE
Use UniFFI to export our logging API (gh-5308)

### PR DESCRIPTION
**Pull request by bendk**
Added a new crate called `rust-logs` that forwards logs to consumers. This does basically the same thing as `rc-log`, but it uses UniFFI. Once our consumer apps swich over to `rust-logs`, we can remove the `rc-log` component.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[ff-android: firefox-android-branch-name]` and/or `[fenix: fenix-branch-name]` to the PR title.

### CODE REVIEW

File: components/support/rust-logs/src/rust_logger.rs
Comment (tarikeshaq) [inline components/support/rust-logs/src/rust_logger.rs:57]: Hmm Do we want to manually set this there? If we do, is it possible that this races with the app calling `set_max_level`? I imagine we **always** want to keep the max level that the app sets, as opposed to the level set by the initializer. Can we add logic here to **not** set the max level if it's already been set by the app? We may be able to read the max level from https://docs.rs/log/latest/log/fn.max_level.html

But I'm not sure what it defaults too, and how we would differentiate between the app setting it, and it being the default value. Another option is to enforce that the app calls this API on the main thread, that's risky though since we use a `RwLock` 🤔 Thoughts?

File: components/support/rust-logs/src/rust_logger.rs
Comment (tarikeshaq) [inline components/support/rust-logs/src/rust_logger.rs:63]: ditto on is this a testing artifact?

File: components/support/rust-logs/src/foreign_logger.rs
Comment (tarikeshaq) [inline components/support/rust-logs/src/foreign_logger.rs:23]: nit: What do you think about removing the explicit lifetime
```suggestion
impl From<&log::Record<'_>> for Record {
```

File: components/support/rust-logs/src/foreign_logger.rs
Comment (tarikeshaq) [inline components/support/rust-logs/src/foreign_logger.rs:12]: nit: What do you think of using a doc comment

**Review by tarikeshaq — CHANGES_REQUESTED** [review]
Overall this makes sense to me as a `uniffi` replacement for rc_log, i love how little code it is!!

I see a bit of complexity in https://github.com/mozilla/application-services/blob/main/components/rc_log/src/android.rs when it comes to stopping the logs, I also see a different approach to getting the tag in its iOS equivalent in https://github.com/mozilla/application-services/blob/f0cf5491ce6f3ccdd80913b9a9a4c3568c6cb2da/components/rc_log/src/ios.rs#L51-L53

We should probably have iOS and android PRs ready before landing this, so we can test to make sure it works properly, and so we don't keep `iOS` hanging like we did for `error_support` 😢 

cc @lougeniaC64 re: iOS

File: components/support/rust-logs/src/rust_logs.udl
Comment (bendk) [reply components/support/rust-logs/src/rust_logs.udl:20]: It comes from the `target` field, which is set to the module name by default.  However, you can manually set it for example [https://docs.rs/log/latest/log/#examples].  It seems pretty weird to manually set it though.

File: components/support/rust-logs/src/rust_logs.udl
Comment (bendk) [reply components/support/rust-logs/src/rust_logs.udl:6]: The idea is that you can pass `None` in to disable log forwarding.  I'll make sure to add a comment saying that.

File: megazords/full/src/lib.rs
Comment (bendk) [reply megazords/full/src/lib.rs:19]: Those comments are a good idea.  Here's the ticket: https://mozilla-hub.atlassian.net/browse/SYNC-3589

File: components/support/rust-logs/src/rust_logger.rs
Comment (bendk) [reply components/support/rust-logs/src/rust_logger.rs:52]: :) pretty sure that was from when I was battling infinite recursion.

**Comment by bendk** [comment]
I made this a new module so we can merge this one early without causing breaking changes to our consumers.  The old module should still work.  Enabling both break things (I think it causes a panic), but we can just avoid doing that.

I have code changes for android, but I forgot to mention them:
 - [Android-components changes](https://github.com/mozilla-mobile/firefox-android/pull/728)
 - [Fenix changes](https://github.com/mozilla-mobile/fenix/pull/28788) (these can actually be merged now).


> I see a bit of complexity in https://github.com/mozilla/application-services/blob/main/components/rc_log/src/android.rs when it comes to stopping the logs.

AFAICT that should be handled by using UniFFI.

> I also see a different approach to getting the tag in its iOS equivalent in ...

Yeah, we could just use `module_name()` instead of `target`.  I'm not sure which one is better or worse.

File: components/support/rust-logs/src/rust_logger.rs
Comment (bendk) [reply components/support/rust-logs/src/rust_logger.rs:55]: Yes, probably the most likely would be an application trying to use both `rust_logs` and `rc_log`, or maybe if one of our components called `log::set_logger()`.  Either way, not something that should happen in practice, so I think panicing is okay.

File: components/support/rust-logs/src/foreign_logger.rs
Comment (bendk) [reply components/support/rust-logs/src/foreign_logger.rs:23]: TIL!

File: components/support/rust-logs/src/rust_logger.rs
Comment (bendk) [reply components/support/rust-logs/src/rust_logger.rs:57]: Good point.  The reason this is here is that Android doesn't set the max level and that results in way too many log entries being emitted.  I'll change it so we remember if the app has set a max level and only set a default if that hasn't happened.

I guess there's still a race if you had 1 thread calling `set_logger` and another calling `set_max_level()`, but that seems like something we can ignore.

File: components/support/rust-logs/src/rust_logger.rs
Comment (bendk) [reply components/support/rust-logs/src/rust_logger.rs:33]: I think we should implement this, but I'm worried about the performance implications of taking the lock.  I'm going to store an atomicbool instead that tracks this to avoid the overhead.

**Comment by bendk** [comment]
Thanks for all the great suggestions.  I think the new commit covers them all, but tell me if I missed any.

File: components/support/rust-logs/src/rust_logs.udl
Comment (mhammond) [reply components/support/rust-logs/src/rust_logs.udl:9]: as a matter of interest, what would block us using procmacros here? Then there's only 1 place the docs could possibly be :)

File: components/support/rust-logs/src/rust_logs.udl
Comment (mhammond) [reply components/support/rust-logs/src/rust_logs.udl:14]: yeah, I love that this is the enum from the `log` crate! Is that a blocker to using procmacros? 😅

**Review by mhammond — COMMENTED** [review]
Just had a quick look, but I agree with Tarik's implication that we need to better understand the fancy things like `std::sync:: mpsc` and threads in the existing code - which I don't :)

**Comment by bendk** [comment]
I'm still not sure what's supposed to be happening `LogAdapterState` and it's threading/mpsc usage, but I don't think it's actually used.  I just deleted that code and a-s still built fine and the a-c tests passed.

File: components/support/rust-logs/src/rust_logs.udl
Comment (bendk) [reply components/support/rust-logs/src/rust_logs.udl:14]: Unfortunately it is currently.  You can't expose a remote type using the proc-macros yet.

File: components/support/rust-logs/src/foreign_logger.rs
Comment (bendk) [reply components/support/rust-logs/src/foreign_logger.rs:20]: `Log` sounds good to me, I'll rename.  I think we would want to wrap the `log` crate as much as possible, but wrapping the `Record` type doesn't seem right to me since it only exposes it's data through methods.  It seems better to send the data over as owned data in a struct.

**Comment by tarikeshaq** [comment]
>  it's threading/mpsc usage,

One way to explain it is it's trying to not block the calling Rust code on the callback, by sending the message to another thread to deal with it and resuming. If that's the case, then I still like this PR, and I would want the calling app to enforce that (i.e, the app can maybe have its callback always trigger a different co-routine to actually log as to not block the thread if it's flushing or something like that)

File: components/support/rust-logs/Cargo.toml
Comment (bendk) [reply components/support/rust-logs/Cargo.toml:2]: Also don't want to bike-shed to much, but what about `rust-log-forwarder`?  That seems to best describe what this crate is responsible for.

**Review by tarikeshaq — APPROVED** [review]
This is looking great! I added one last comment about testing the max level setting, but no need to block a review from me after that!

I think it'll be important however to get Android eyes on the Android PRs before we land this and communicate the change in the threading model (we no longer launch a logger thread, but now the Rust code will block on the callback)

Approving to remove the "Request Changes"!

**Comment by bendk** [comment]
I thought this would be so simple, but I spent most of the day looking into 2 very weird issues:

- AFAICT, the `android.rs` module is never used.  It's wrapped in `cfg(any(test, os = "android", feature = "force_android"))`, but:
  - When I use the local auto-publish it's not used.  I can delete the entire module and still build and test
  - I don't see any code that sets `force_android`.
  - Is `os=android` the right check?  I believe `target_os=android` might be what they wanted, but I don't think a plain `os` is set as a config value, and if it was wouldn't it reflect the build system not the target?
- I did some performance testing and it seems the new code is 10x slower than the old.  I tested by sending 1 million log messages and timing it in a unit test.  The old code took around 1.8 second, the new code around 18.  I tried to narrow this down, but couldn't figure out exactly what the issue was.  I believe it's related to making the actual call rather than packing the data into a RustBuffer, but I could be wrong about that.

Part of me wants to just ignore the performance issue, but maybe we should dig deeper into it since we want to use callbacks for more things.  What do others think?

**Comment by tarikeshaq** [comment]
> Part of me wants to just ignore the performance issue, but maybe we should dig deeper into it since we want to use callbacks for more things. What do others think?

+1 to looking into it more. I'm not sure if there is a good way to profile it to see why it's slower, if the `android.rs` code did in fact run, then I suspect the offloading to another thread could be a good candidate for the performance

I'd be hesitant to ignore the performance issue because we log a decent bit when we sync, I can imagine a history sync starting to take a noticeably longer time because we log a lot as we apply records, etc. That said, some benchmarking of sync can convince me otherwise (i.e can we show that logging is very far from being the bottleneck, even times 10)

**Comment by tarikeshaq** [comment]
`Is os=android the right check? I believe target_os=android`

Heh I think you're right. I think the whole time we've been using the iOS logger for Android 😨 we need to confirm, but `os` isn't a valid `cfg` value. Tried messing around with changing things to `os = "macos"` in case it's deprecated or something and the compiler isn't compiling things

**Comment by bendk** [comment]
I did some more performance testing and I think I understand what's going on, although it's not totally clear what the call should be.  My main test was to add a ruts function that made 1,000,000 log calls, call it in a Kotlin unit test and time the entire operation.  The current code took about 1800ms.  I'm pretty sure that it actually made the calls, not just scheduled them to a thread.  In addition to the fact that it seems like the `cfg` selector is broken, I reduced the `mpsc` size to 0 and the time was the same.  The code from this branch took about 18,000ms (10x slower).

Note: I mostly trust my work but I'm not totally sure about it.   I think we can have enough confidence to make our decision, but don't take anything here as completely reliable.

The time for the actual Kotlin logging calls seems small.  I can remove them and the total time goes to about 17,000ms.

A lot of time is spent in the [kotlin generated code that lifts arguments, makes the call, and lowers returns](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt#L27-L82).  If I short-circuit that, the time goes down to about 6,000ms.
  - This is after I removed the code from the `log()` method, so the inner method shouldn't be contributing much time here.  I think it's just the code wrapping it.
  - I don't think buffer packing/unpacking is the issue here.  If I remove the `message` and `tag` fields from `Record`, which are most of the data, the time doesn't change.  My feeling is the time is mostly taken in JNA code that's creating/reading/writing the structs.

I think the other main time is simply invoking the C function -- specifically the work JNA does to convert the arguments/return values.  After short-circuiting the invoke() method, I tried removing all arguments from [`ForeignCallback`](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_core/src/ffi/foreigncallbacks.rs#L142-L147).  This reduced the time to about 700ms.  Skipping the call from Rust to the foreign callback reduced the time down to 100ms.

So I think the difference in performance is mostly JNA stuff.  The old function only inputted ints and c strings.  Now we're inputting, returning, and creating RustBuffer and other structs and that's slowing things down.  I think to get any more accurate I would need to profile the JNA code, but I don't think I'm up for that.

Maybe UniFFI can try to optimize how it calls callbacks better, but I'm not sure exactly how.  If we assume that the performance isn't going to improve, are we okay with it?

These numbers indicate that it takes about 18ms to log 1,000 items on my machine.  That seems alright, but actual android apps are probably going to be slower, but I don't know how much slower.  If a slower phone took 200ms to log 1,000 items that doesn't seem too bad to me.  But anything slower feels bad.  What do others think about this?

**Comment by tarikeshaq** [comment]
Thank you for the in-depth analysis @bendk 🚀 🎉 
> although it's not totally clear what the call should be.

We can get an answer by getting mobile eyes on the Fenix PR - if we add some android folks I'm sure they'll be able to tell us if the performance regression is one we can live with.

> 18ms to log 1,000 items on my machine

To my naive eyes seems more than fine, especially when I start to remember that the android logger probably is smart enough to store things in memory and flush when it needs to, etc


I was mainly concerned about **The Rust code blocking on the log, and causing it to slow down other Rust operations** but given your analysis, it shouldn't be a problem. 

I'd like us to get an Android +1 before we land this, but I wouldn't mind landing it either (as long as we get that +1 on the fenix/android-components PR anyways)

**Comment by bendk** [comment]
The more I think about this one, the less I like taking a performance hit on Android for this.  What if we:
  - Adding the new logging crate and kept the old `rc_log` crate
  - Kept Android on `rc_log`
  - Switched iOS to `rust_log_forwarder`.  Since JNA seems to be the issue here, I don't think we'll have the same performance issues on iOS.
  - Use `rust_log_forwarder` for Desktop.

We'd have to maintain two crates, including one that's messy code-wise, but this logging code doesn't change much.  I think I'd choose that over having to figure out the performance question with the Android team.

**Comment by tarikeshaq** [comment]
> Kept Android on rc_log

Will there be a medium-long term plan to move Android away from `rc_log`? I don't like having two libraries for logging, one for Android and the other for iOS 😞 

It would be a slight shame to not ship to android before analyzing if there would even be a performance impact. I see https://github.com/mozilla-mobile/fenix#performance-build-variants has some details on benchmarking, maybe we can also ask the mobile team directly too if they have any resources (and we can do the testing!)


> Switched iOS to rust_log_forwarder

I'd also want to do some simple benchmarking here to verify, I'll free up some time later tonight and run similar benchmarks that you did for android (since I have a mac)


I do agree that if the answer to the performance testing will require a lot of work, then we can ship to android later once we have the answers


**None of this is blocking the landing of this PR, so feel free to land and we can figure out how to add it to the apps in a followup ticket (or in the tickets in the apps themselves)**

Thoughts? I'd also want to hear @mhammond's opinion here

**Comment by bendk** [comment]
> Will there be a medium-long term plan to move Android away from `rc_log`? I don't like having two libraries for logging, one for Android and the other for iOS disappointed

Agreed.  If it was just us making the performance decision I would definitely be for it.  But asking for the Android team to assist us on this seems like a bigger cost.

Long-term, I'd love to look at UniFFI and see if we can speed up the performance or maybe give applications a way to write some code by hand to speed things up.

**Comment by mhammond** [comment]
> Thoughts? I'd also want to hear @mhammond's opinion here

I chatted with @bendk about this, and to summarize:
* At face value the performance doesn't seem that bad, but I'm slightly concerned by a kind of "death by a thousand cuts". Eg, I'd love to expose in the "secret debug menu" a way for users to enable trace logging which would be scary on underpowered devices which are already struggling to sync (which is presumably why they might want to enable trace logging in the first place)
* I don't think we have a good understanding of why the old code is so much faster. It appears to be doing the same amount of work roughly. It would be great to better understand that and see whether uniffi can help us here (eg, we discussed that maybe using 3 args instead of a struct might explain it, but that's not clear either - the old code copies the strings into a null terminated buffer and frees them, which almost seems like is *more* work than the rust buffer does?)
* Because we already have a working system for Android, and a *partially* working story for desktop (webext-storage can log, it's just tabs which can't, so an alternative model there might be to lean into the xpcom solution), I'm mildly against landing this new code until we are committed to it - there's no suggestion landing it earlier is easier than landing it later, but landing it later makes it far more likely we will actually do whatever followups we identify.
* Even if we can't get the full perf story, it might still be a reasonable decision to commit to this - but let's see what else we can discover first.

**Comment by bendk** [comment]
I took a closer look at this today.  The best way I could figure out how to report the results was a [set of commits](https://github.com/bendk/application-services/compare/uniffi-logging...bendk:application-services:uniffi-logging-perf?expand=1).  If you go through those 1 by 1 you can see different performance optimizations and their effect.

TLDR: I think with some basic optimizations, we can go from being ~10x than the hand-written FFI to ~1.8x slower.  This seems fine to me and I don't think it should cause other teams to hesitate on using UniFFI.  We could probably increase this more, but this feels like a good cutoff for now.

How do others feel about landing this now, landing the performance changes to UniFFI, then landing the change to firefox-android to start using the new code.

Unfortunately, this changes the ABI for callback interfaces, so I guess we'll need a breaking UniFFI bump.  I was hoping to get at least one patch release in, but oh-well.

**Comment by mhammond** [comment]
> TLDR: I think with some basic optimizations, we can go from being ~10x slower than the hand-written FFI to ~1.8x slower.  This seems fine to me and I don't think it should cause other teams to hesitate on using UniFFI

That's fantastic, and I agree 100%.

> Unfortunately, this changes the ABI for callback interfaces, so I guess we'll need a breaking UniFFI bump. I was hoping to get at least one patch release in, but oh-well.

heh - I'm sure there will be plenty of patch releases in the future though :)

The passing the RustBuffer as 2 args is interesting - do you think that can be used anywhere else?

(Regardless, this is great, thanks!)

**Comment by bendk** [comment]
> The passing the RustBuffer as 2 args is interesting - do you think that can be used anywhere else?

I think it could be used for regular calls as well.  That could even result in a bigger optimization, since we only pass 1 argument buffer for callbacks, but regular calls could have multiple rust buffers involved.

**Pull request merged**
